### PR TITLE
dstream fixes and enhancements (master)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: lib/core/include/stream_factory_utility.hpp [hunk 1]
```diff
@@ -128,7 +128,61 @@ namespace irods::experimental::io
...
+    ///                       types are supported:
+    ///                           - irods::experimental::io::replica_number
+    ///                           - irods::experimental::io::leaf_resource_name
+    ///                           - irods::experimental::io::root_resource_name
+    ///                       \endparblock
+    ///
+    /// \return A new factory function.
+    ///
+    /// \since 4.2.9
+    template <typename ReplicaIdType>
```
Comment (korydraughn) [inline lib/core/include/stream_factory_utility.hpp:154]: I do not like this name. I'd like the name to tell the user that this type is for identifying a specific replica.

Perhaps:
- `ReplicaIdentifyingInfo` (eww)
- `TargetReplicaType`
- Something else?

Comment (trel) [reply lib/core/include/stream_factory_utility.hpp:154]: `ReplicaIdentifier`?

File: unit_tests/src/unit_test_utils.hpp [hunk 2]
```diff
@@ -116,13 +115,13 @@ namespace unit_test_utils
         }
 
         for (std::size_t i = 0; i < _size_in_bytes; ++i) {
-            out << _ch;
+            out.write(&_ch, 1);
```
Comment (trel) [inline unit_tests/src/unit_test_utils.hpp:118]: just curious... why is it '1' here?

https://www.cplusplus.com/reference/ostream/ostream/write/

Comment (korydraughn) [reply unit_tests/src/unit_test_utils.hpp:118]: We are writing one byte to the stream.

File: lib/core/include/stream_factory_utility.hpp (hunk 1 above)
Comment (korydraughn) [reply lib/core/include/stream_factory_utility.hpp:154]: I'll use that.

File: unit_tests/src/unit_test_utils.hpp (hunk 2 above)
Comment (trel) [reply unit_tests/src/unit_test_utils.hpp:118]: oh, dang, a for loop.   good lord.   carry on.

**Review by alanking — APPROVED** [review]

File: lib/core/include/stream_factory_utility.hpp (hunk 1 above)
Comment (korydraughn) [reply lib/core/include/stream_factory_utility.hpp]: done

File: lib/core/include/stream_factory_utility.hpp [hunk 3]
```diff
@@ -128,7 +128,61 @@ namespace irods::experimental::io
...
+    ///
+    /// \return A new factory function.
+    ///
+    /// \since 4.2.9
+    template <typename ReplicaIdentifier>
+    auto make_dstream_factory(irods::connection_pool& _conn_pool,
+                              std::string _logical_path,
+                              ReplicaIdentifier&& _replica_id)
+    {
+        return [&_conn_pool, p = std::move(_logical_path), repl_id = std::forward<ReplicaIdentifier>(_replica_id)]
```
Comment (korydraughn) [inline lib/core/include/stream_factory_utility.hpp:159]: should `std::forward` be used here or does a normal capture by reference maintain the type information for forwarding later.

Comment (korydraughn) [reply lib/core/include/stream_factory_utility.hpp:159]: through experimentation against clang and gcc, capturing by reference is the way to go.

**Comment by trel** [comment]
these look good to me - # when ready

**Comment by korydraughn** [comment]
Added #

**Pull request merged**